File: code/432.cpp
#include <bits/stdc++.h>
#define INF 2000000000
using namespace std;
typedef long long ll;
int read(){
	int f = 1, x = 0;
	char c = getchar();
	while(c < '0' || c > '9'){if(c == '-') f = -f; c = getchar();}
	while(c >= '0' && c <= '9')x = x * 10 + c - '0', c = getchar();
	return f * x; 
}
// 双向链表都能忘，爬
class AllOne {
    unordered_map<int, unordered_set<string>> val_to_key;
    unordered_map<string, int> key_to_val;
    unordered_map<int, int> left, right;
public:
    /** Initialize your data structure here. */
    AllOne() {
        left[0] = right[0];
    }
    
    /** Inserts a new key <Key> with value 1. Or increments an existing key by 1. */
    void inc(string key) {
        if (key_to_val.count(key)){
            int v = key_to_val[key];
            ++key_to_val[key];
            val_to_key[v + 1].insert(key);
            if (val_to_key[v + 1].size() == 1){
                int nxt = right[v];
                left[nxt] = v + 1, right[v + 1] = nxt;
                left[v + 1] = v, right[v] = v + 1;
            }
            val_to_key[v].erase(key);
            if (val_to_key[v].empty()){
                int prev = left[v];
                left[v + 1] = prev, right[prev] = v + 1;
            }
        }else {
            key_to_val[key] = 1;
            val_to_key[1].insert(key);
            if (val_to_key[1].size() == 1){
                int nxt = right[0];
                left[nxt] = 1, right[1] = nxt;
                left[1] = 0, right[0] = 1;
            }
        }
    }
    
    /** Decrements an existing key by 1. If Key's value is 1, remove it from the data structure. */
    void dec(string key) {
        if (!key_to_val.count(key)) return ;
        int v = key_to_val[key];
        if (v > 1){
            --key_to_val[key];
            val_to_key[v - 1].insert(key);
            if (val_to_key[v - 1].size() == 1){
                int prev = left[v];
                left[v - 1] = prev, right[v - 1] = v;
                left[v] = v - 1, right[prev] = v - 1;
            }
            val_to_key[v].erase(key);
            if (val_to_key[v].empty()){
                int nxt = right[v];
                left[nxt] = v - 1, right[v - 1] = nxt;
            }
        }else {
            key_to_val.erase(key);
            val_to_key[1].erase(key);
            if (val_to_key[1].empty()){
                int nxt = right[1];
                left[nxt] = 0, right[0] = nxt;
            }
        }
    }
    
    /** Returns one of the keys with maximal value. */
    string getMinKey() {
        int mini = right[0];
        if (!mini) return string("");
        for (auto& s: val_to_key[mini])
            return s;
        return string("");
    }
    
    /** Returns one of the keys with Minimal value. */
    string getMaxKey() {
        int maxi = left[0];
        if (!maxi) return string("");
        for (auto& s: val_to_key[maxi])
            return s;
        return string("");
    }
};

/**
 * Your AllOne object will be instantiated and called as such:
 * AllOne* obj = new AllOne();
 * obj->inc(key);
 * obj->dec(key);
 * string param_3 = obj->getMaxKey();
 * string param_4 = obj->getMinKey();
 */
AllOne sol;
void init(){
    
}
void solve(){
	// sol.convert();
}
int main(){
	init();
	solve();
	return 0;
}

File: code/432.py
class AllOne:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.key_to_val = dict()
        self.val_to_key = dict()
        self.mini = -1
        self.maxi = -1

    def inc(self, key: str) -> None:
        """
        Inserts a new key <Key> with value 1. Or increments an existing key by 1.
        """
        if key in self.key_to_val:
            # 更新键值对
            v = self.key_to_val[key]
            self.key_to_val[key] += 1
            self.val_to_key[v].remove(key)
            if v + 1 not in self.val_to_key:
                self.val_to_key[v + 1] = set()  
            self.val_to_key[v + 1].add(key)
            
            # 更新最小值，此时最小值必存在
            if self.mini == v and not self.val_to_key[v]:
                self.mini += 1
            # 更新最大值，此时最大值必存在
            self.maxi = max(self.maxi, v + 1)
        else:
            # 插入键值对
            self.key_to_val[key] = 1
            if 1 not in self.val_to_key:
                # 此时初始化
                self.val_to_key[1] = set()
                self.val_to_key[1].add(key)
                self.mini = 1
                self.maxi = 1
                return 
            self.val_to_key[1].add(key)

            # 更新最小值，此时最小值必为 1
            self.mini = 1
            # 更新最大值
            self.maxi = max(1, self.maxi)

    def dec(self, key: str) -> None:
        """
        Decrements an existing key by 1. If Key's value is 1, remove it from the data structure.
        """
        v = self.key_to_val.get(key, -1)
        if v < 0:
            return 
        # 删除这个键
        self.key_to_val.pop(key)
        self.val_to_key[v].remove(key)
        
        if v == 1:
            # 需要更新最大值和最小值
            



        else:

        

    def getMaxKey(self) -> str:
        """
        Returns one of the keys with maximal value.
        """
        if self.maxi < 0: return ""
        for s in self.val_to_key[self.maxi]:
            # 遇到就直接返回
            return s
        return ""

    def getMinKey(self) -> str:
        """
        Returns one of the keys with Minimal value.
        """
        if self.mini < 0: return ""
        for s in self.val_to_key[self.mini]:
            # 遇到就直接返回
            return s
        return ""
        


# Your AllOne object will be instantiated and called as such:
# obj = AllOne()
# obj.inc(key)
# obj.dec(key)
# param_3 = obj.getMaxKey()
# param_4 = obj.getMinKey()